injectproxy/enforce.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package injectproxy
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+type Enforcer struct {
+	labelMatchers map[string]*labels.Matcher
+}
+
+func NewEnforcer(ms ...*labels.Matcher) *Enforcer {
+	entries := make(map[string]*labels.Matcher)
+
+	for _, matcher := range ms {
+		entries[matcher.Name] = matcher
+	}
+
+	return &Enforcer{
+		labelMatchers: entries,
+	}
+}
+
+// EnforceNode walks the given node recursively
+// and enforces the given label enforcer on it.
+//
+// Whenever a parser.MatrixSelector or parser.VectorSelector AST node is found,
+// their label enforcer are being potentially modified.
```

Comment:
tiny grammar nit: we need correspondence between `are` and the singular `label enforcer`

Reply:
Is it the label enforcer of the node that is being modified? Or the label matcher of the node?

Reply:
@squat ah, thank you for grammar corrections :pray: it is the latter, the label matcher of node. The enforcer is immutable it is "configuration data" if that makes sense.